collector/analyzer/tcpconnectanalyzer/internal/connect_monitor.go

```diff
@@ -0,0 +1,327 @@
+		ce.Write(
+			zap.String("ConnKey", connKey.String()),
+			zap.Int64("retValue", retValueInt),
+		)
+	}
+
+	connStats, ok := c.connMap[connKey]
+	if !ok {
+		// Maybe the connStats have been closed by tcp_set_state_from_established event.
+		// We don't care about it.
+		return nil, nil
+	}
+	// "connect_exit" comes to analyzer after "tcp_connect"
+	connStats.EndTimestamp = event.Timestamp
+	connStats.pid = event.GetPid()
+	connStats.ContainerId = event.GetContainerId()
+	var eventType EventType
+	if retValueInt == 0 {
+		eventType = connectExitSyscallSuccess
+	} else if isNotErrorReturnCode(retValueInt) {
+		eventType = connectExitSyscallNotConcern
+	} else {
+		eventType = connectExitSyscallFailure
+		connStats.Code = int(retValueInt)
+	}
+	return connStats.StateMachine.ReceiveEvent(eventType, c.connMap)
+}
+
+func (c *ConnectMonitor) ReadSendRequestSyscall(event *model.KindlingEvent) (*ConnectionStats, error) {
+	// The events without sip/sport/dip/dport have been filtered outside this method.
+	connKey := ConnKey{
+		SrcIP:   event.GetSip(),
+		SrcPort: event.GetSport(),
+		DstIP:   event.GetDip(),
+		DstPort: event.GetDport(),
+	}
+	if ce := c.logger.Check(zapcore.DebugLevel, "Receive sendRequestSyscall event:"); ce != nil {
+		ce.Write(
+			zap.String("ConnKey", connKey.String()),
+			zap.String("eventName", event.Name),
+		)
+	}
+
+	connStats, ok := c.connMap[connKey]
+	if !ok {
+		return nil, nil
+	}
+	connStats.pid = event.GetPid()
+	connStats.ContainerId = event.GetContainerId()
+	return connStats.StateMachine.ReceiveEvent(sendRequestSyscall, c.connMap)
+}
+
+func isNotErrorReturnCode(code int64) bool {
+	return code == einprogress || code == eintr || code == eisconn || code == ealready
+}
+
+func (c *ConnectMonitor) ReadInTcpConnect(event *model.KindlingEvent) (*ConnectionStats, error) {
+	connKey, err := getConnKeyForTcpConnect(event)
+	if err != nil {
+		return nil, err
+	}
+	retValue := event.GetUserAttribute("retval")
+	if retValue == nil {
+		return nil, fmt.Errorf("retval of tcp_connect is nil")
+	}
+	retValueInt := retValue.GetUintValue()
+
+	if ce := c.logger.Check(zapcore.DebugLevel, "Receive tcp_connect event:"); ce != nil {
+		ce.Write(
+			zap.String("ConnKey", connKey.String()),
+			zap.Uint64("retValue", retValueInt),
+		)
+	}
+
+	var eventType EventType
+	if retValueInt == 0 {
+		eventType = tcpConnectNoError
+	} else {
+		eventType = tcpConnectError
+	}
+
+	connStats, ok := c.connMap[connKey]
+	if !ok {
+		// "tcp_connect" comes to analyzer before "connect_exit"
+		connStats = &ConnectionStats{
+			ConnKey:          connKey,
+			InitialTimestamp: event.Timestamp,
+			EndTimestamp:     event.Timestamp,
+			Code:             int(retValueInt),
+		}
+		connStats.StateMachine = NewStateMachine(Inprogress, c.statesResource, connStats)
+		c.connMap[connKey] = connStats
+	} else {
+		// Not possible to enter this branch
+		connStats.EndTimestamp = event.Timestamp
+		connStats.Code = int(retValueInt)
+	}
```

Comment:
I thought that might happen when kindling missed the `ConnectExitSyscall` and the process restarted connecting before expired.  Could this `ConnKey` be the same as the old one?

Reply:
Only `tcp_connect` event could create a new `connStats`, and there is only one `tcp_connect` event for each TCP socket. I haven't seen there are two `tcp_connect` events for the same socket, thus I say this branch is impossible to be entered.